daliuge-engine/dlg/manager/node_manager.py

```diff
@@ -111,6 +111,108 @@ def _load(obj, callable_attr):
         )
     return obj
 
+class NodeManagerDropRunner(DropRunner):
+    @abc.abstractmethod
+    def start(self, rpc_endpoint):
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def close(self):
+        raise NotImplementedError
+
+
+class NodeManagerThreadDropRunner(NodeManagerDropRunner):
+    def __init__(self, max_workers: int):
+        self._max_workers = max_workers
+        self._thread_pool: ThreadPoolExecutor | None = None
+
+    def start(self, _rpc_endpoint):
+        logger.info("Initializing thread pool with %d workers", self._max_workers)
+        self._thread_pool = ThreadPoolExecutor(max_workers=self._max_workers)
+
+    def run_drop(self, app_drop: AppDROP) -> Future:
+        return self._thread_pool.submit(app_drop.run)
+
+    def close(self):
+        self._thread_pool.shutdown(wait=True)
+
+
+class NodeManagerProcessDropRunner(NodeManagerDropRunner):
+    # Process isolated properties - should only be accessed in @classmethods
+    # to ensure that they are global to a single process only
+    _rpc_client: typing.Optional[rpc.RPCClient]
+    _rpc_endpoint: typing.Tuple[str, int]
+
+    def __init__(self, max_workers: int):
+        self._max_workers = max_workers
+        self._process_pool: typing.Optional[ProcessPoolExecutor] = None
+
+    def start(self, rpc_endpoint):
+        logger.info("Initializing process pool with %d workers", self._max_workers)
+
+        self._process_pool = ProcessPoolExecutor(
+            max_workers=self._max_workers,
+            initializer=NodeManagerProcessDropRunner._setup_process,
+            initargs=(rpc_endpoint,),
+        )
+
+    @classmethod
+    def _setup_process(cls, rpc_endpoint):
+        signal.signal(signal.SIGTERM, signal.SIG_IGN)
+        signal.signal(signal.SIGINT, signal.SIG_IGN)
+
+        cls._rpc_endpoint = rpc_endpoint
+        cls._rpc_client = rpc.RPCClient()
+
+        # We'll just let this be cleaned up when the process terminates
+        # (instead of the normal `shutdown()`)
+        cls._rpc_client.start()
+
+    def run_drop(self, app_drop: AppDROP):
+        inputs_proxy_info, outputs_proxy_info = NodeManagerProcessDropRunner._get_proxy_infos(app_drop)
+
+        # MP Queues pickle on a background thread - we need to ensure that we don't
```

Comment:
Did you experience any issues if *not* doing this? My understanding of this background thread was that it *flushes* on the background, but pickling would happen synchronously, meaning that in principle there shouldn't be a problem in not making a copy.

Reply:
Yep :/ Previously we would clear the inputs and outputs in here, submit it to the pool, then revert them back to their original values before returning to the caller. I found I'd get issues where errors would be thrown for unpicklable values (which confused me for ages!) which I eventually narrowed down to being input/output values that I'd thought wouldn't have been there! To add to it, when I was debugging it didn't show the issue due to the pickling finishing before I stepped out of the function (and the inputs/outputs got set back).

We could also avoid this by not setting the input/output back again, but I wasn't familiar enough with the code base to be confident that wouldn't break other things. A copy seemed safest.

Looking at it again now, the handling looks a little more complicated that I remember (see [here](https://github.com/python/cpython/blob/v3.10.13/Lib/concurrent/futures/process.py#L717)), but I definitely hit issues, and this definitely fixed it.

Reply:
OK, thanks for that explanation! The wording in the https://docs.python.org/3/library/multiprocessing.html#pipes-and-queues isn't crystal clear TBH, so I always assumed things got pickled in the foreground, good to know that's actually not the case!